module/persistent.lua

```diff
@@ -0,0 +1,192 @@
+local awful = require("awful")
+local gears = require("gears")
+
+local capi =
+{
+    awesome = awesome,
+    root = root,
+    screen = screen,
+    client = client
+}
+local ipairs = ipairs
+local table = table
+local tonumber = tonumber
+local tostring = tostring
+
+local persistent = { mt = {} }
+local instance = nil
+
+local function get_xproperty(name, type)
+    capi.awesome.register_xproperty(name, type)
+    return capi.awesome.get_xproperty(name)
+end
+
+local function set_xproperty(name, type, value)
+    capi.awesome.register_xproperty(name, type)
+    capi.awesome.set_xproperty(name, value)
+end
+
+local function client_get_xproperty(client, name, type)
+    local xprop = "bling.client." .. name
+    capi.awesome.register_xproperty(xprop, type)
+    return client:get_xproperty(xprop)
+end
+
+local function client_set_xproperty(client, name, type, value)
+    local xprop = "bling.client." .. name
+    capi.awesome.register_xproperty(xprop, type)
+    client:set_xproperty(xprop, value)
+end
+
+local function save()
+    set_xproperty("tag_index", "number", awful.screen.focused().selected_tag.index)
+
+    for index, entry in ipairs(capi.root.tags()) do
+        local master_width_property = "tag_" .. entry.index .. "_master_width_factor"
+        set_xproperty(master_width_property, "string", tostring(entry.master_width_factor))
+
+        local layout_property = "tag_" .. entry.index .. "_layout"
+        set_xproperty(layout_property, "number", awful.layout.get_tag_layout_index(entry))
+    end
+
+    for index, client in ipairs(capi.client.get()) do
```

Comment:
I think you can save some brain cells in this for loop with something like:

```lua
local properties = { "hidden", "minimized", } -- add more in the table
for _, property in ipairs(properties) do
    client_set_xproperty(client, property, type(client[property]), client[property])
end
```

I think the only exception that wouldn't work with that are properties like `screen` where you save `screen.index`.